sdks/python/apache_beam/runners/runner.py

```diff
@@ -119,6 +119,33 @@ def visit_transform(self, transform_node):
           logging.error('Error while visiting %s', transform_node.full_label)
           raise
 
+    class GroupByKeyInputVisitor(PipelineVisitor):
+      """A visitor that replaces `Any` element type for input `PCollection` of
+      a `GroupByKey` with a `KV` type.
+
+      TODO(BEAM-115): Once Python SDk is compatible with the new Runner API,
+      we could directly replace the coder instead of mutating the element type.
+      """
+      def visit_transform(self, transform_node):
+        # Imported here to avoid circular dependencies.
+        # pylint: disable=wrong-import-order, wrong-import-position
+        from apache_beam import GroupByKey
+        from apache_beam import typehints
+        if isinstance(transform_node.transform, GroupByKey):
+          pcoll = transform_node.inputs[0]
+          input_type = pcoll.element_type
+          if not isinstance(input_type, typehints.TupleHint.TupleConstraint):
+            if isinstance(input_type, typehints.AnyTypeConstraint):
+              # `Any` type needs to be replaced with a KV[Any, Any] to
+              # force a KV coder as the main output coder for the pcollection
+              # preceding a GroupByKey.
+              pcoll.element_type = typehints.KV[typehints.Any, typehints.Any]
+            else:
```

Comment:
This is a bit to strong, e.g. Union[KV[str, int], KV[str, float]] could be fine, but that's probably beyond the scope of this CL (though perhaps worth dropping a TODO).